tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,447 @@
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/gpu/horizontal_fusion.h"
+
+#include <algorithm>
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/types/span.h"
+#include "tensorflow/compiler/xla/layout_util.h"
+#include "tensorflow/compiler/xla/service/hlo_creation_utils.h"
+#include "tensorflow/core/util/env_var.h"
+
+namespace xla {
+namespace gpu {
+
+namespace {
+
+absl::InlinedVector<HloInstruction*, 2> GetOutputsOfFusion(
+    const HloInstruction& insn) {
+  CHECK(insn.opcode() == HloOpcode::kFusion);
+  auto root = insn.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return {root};
+  } else {
+    return root->operands();
+  }
+}
+
+// Return the number of outputs of the fused computation.
+size_t GetOutputSizeOfFusion(const HloInstruction& insn) {
+  CHECK(insn.opcode() == HloOpcode::kFusion);
+  auto root = insn.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return 1;
+  } else {
+    return ShapeUtil::TupleElementCount(root->shape());
+  }
+}
+
+class HorizontalFusionImpl {
+ public:
+  HorizontalFusionImpl(HloComputation* computation)
+      : computation_(computation) {}
+
+  ~HorizontalFusionImpl() {}
+
+  StatusOr<bool> Run();
+
+ private:
+  Status Fuse(absl::Span<HloInstruction*> fused_fusion_insns);
+
+  Status CreateFusedComputation(
+      absl::Span<HloInstruction*> fused_fusion_insns,
+      std::unique_ptr<HloComputation>* uniq_computation,
+      std::vector<HloInstruction*>* bound_operands);
+
+  // FusionCandidates collects profitable candidates for a given consumer
+  // instruction. GetNextSpanOfFusions() can then be iteratively invoked to
+  // acquire the next set of fusion candidates based on some heuristics.
+  class FusionCandidates {
+   public:
+    FusionCandidates(HloInstruction* consumer) : fusion_insns_(), pos_(0) {
+      Initialize(consumer);
+    }
+
+    // Get a span of fusions to be fused.
+    absl::Span<HloInstruction*> GetNextSpanOfFusions();
+
+   private:
+    void Initialize(HloInstruction*);
+
+    std::vector<HloInstruction*> fusion_insns_;
+    // pos_ points to the start position of next span.
+    size_t pos_;
+  };
+
+  HloComputation* computation_;
+};  // HorizontalFusionImpl
+
+void HorizontalFusionImpl::FusionCandidates::Initialize(
+    HloInstruction* consumer) {
+  // First, find out all fusion instructions. We will filter out
+  // unsupported/non-profitable cases below.
+  absl::flat_hash_set<HloInstruction*> fusion_insns;
+  for (auto opnd : consumer->operands()) {
+    auto predecessor = opnd->LatestNonGteAncestor();
+    if (predecessor->opcode() == HloOpcode::kFusion) {
+      fusion_insns.insert(predecessor);
+    }
+  }
+
+
+  auto is_supported_fusion = [&](const HloInstruction& insn) -> bool {
+    // Support only kLoop fusion.
+    if (!insn.IsLoopFusion()) {
+      return false;
+    }
+
+    return true;
```

Comment:
> The only requirement for the computation itself should be that the fused computations have the same number of outputs.

Interesting, which non-element-wise ops that satisfy this property? I can think of bitcast, reshape, reverse, and transpose. Is that right? Do reshapes and transposes make sense for horizontal fusion?

Reply:
They are all fine as long as they are in the kLoop fusion.

In other words, we rely on _all_ the concatenated outputs in the (slice) input fusion to have the same shape, so that the kernel generation can use this shape as kernel launch dim. To ensure the concatenated outputs having the same shape, we currently rely on two requirements--(a) the fused computations are kLoop fusion and (b) the fused computations have the same number of outputs.

Then, it guarantees that the following two concatenated outputs have the same shapes, assuming we have two fused kLoop computations, each having two outputs (`o1` and `o2`):
`concatenate(computation1.o1, computation2.o1)`
and `concatenate(computation1.o2, computation2.o2)`